File: src/common/py/api/project/project_logbook_events.py
from ...core.messaging import (
    Message,
    Event,
)
from ...core.messaging.composers import (
    MessageBuilder,
    EventComposer,
)
from ...data.entities.project import Project, ProjectID


@Message.define("event/project/logbook/update")
class ProjectLogbookUpdateEvent(Event):
    """
    Sent whenever the logbook of a project has changed.

    Args:
        project_id: The ID of the updated project.
        logbook: The new logbook contents.
    """

    project_id: ProjectID

    logbook: Project.Logbook

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        logbook: Project.Logbook,
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ProjectLogbookUpdateEvent,
            chain,
            project_id=project_id,
            logbook=logbook,
        )

File: src/common/web/api/project/ProjectLogbookEvents.ts
import { Type } from "class-transformer";

import { EventComposer } from "../../core/messaging/composers/EventComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Event } from "../../core/messaging/Event";
import { Message } from "../../core/messaging/Message";
import { ProjectLogbook } from "../../data/entities/project/logbook/ProjectLogbook";
import { type ProjectID } from "../../data/entities/project/Project";

/**
 * Sent whenever the logbook of a project has changed.
 *
 * @param project_id - The ID of the updated project.
 * @param logbook - The new logbook contents.
 */
@Message.define("event/project/logbook/update")
export class ProjectLogbookUpdateEvent extends Event {
    public readonly project_id: ProjectID = 0;

    // @ts-ignore
    @Type(() => ProjectLogbook)
    public readonly logbook: ProjectLogbook = new ProjectLogbook();

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        projectID: ProjectID,
        logbook: ProjectLogbook,
        chain: Message | null = null,
    ): EventComposer<ProjectLogbookUpdateEvent> {
        return messageBuilder.buildEvent(ProjectLogbookUpdateEvent, { project_id: projectID, logbook: logbook }, chain);
    }
}
